### DISCUSSION TITLE
Is it possible to solve transient anisotropic heat transfer equation using py-pde?

### DISCUSSION THREAD

**Discussion by louis10643** [zwicker-group/py-pde#366] [Q&A] [answered]
Hi,

Is it possible to solve transient anisotropic heat transfer equation using py-pde? 
![image](https://user-images.githubusercontent.com/29681897/220331481-332a4c34-59bf-4cb9-b6f3-bdbc435375d5.png)

where kx, ky, kz are all different.

I tried to use diffusion module but its D only takes constant as parameter.

**Comment by david-zwicker** [comment, marked as answer]
This is definitely possible, albeit not with the simple `DiffusionPDE` class. Instead, you might want to use the powerful [`PDE` class](https://py-pde.readthedocs.io/en/latest/packages/pde.pdes.pde.html#pde.pdes.pde.PDE), similar to [this example](https://github.com/zwicker-group/py-pde/blob/master/examples/pde_heterogeneous_diffusion.py).

**Reply by louis10643** [reply]
Thx for the reply! I did some research and "almost" figure out how to do it. If I don't have any internal heat source (qg), the follow code works fine. My question is how to add this internal heat source term (qg)?
Here is my code:

```python
import pde
from pde import CartesianGrid, MemoryStorage, PlotTracker, PDE,ScalarField, plot_kymograph
import functools

#register operators
from pde.grids.operators.cartesian import _make_derivative
from pde.grids.cartesian import CartesianGridBase

make_gradient_x = functools.partial(_make_derivative, axis=0)
CartesianGridBase.register_operator('gradient_x', make_gradient_x)

make_gradient_y = functools.partial(_make_derivative, axis=1)
CartesianGridBase.register_operator('gradient_y', make_gradient_y)

make_gradient_z = functools.partial(_make_derivative, axis=2)
CartesianGridBase.register_operator('gradient_z', make_gradient_z)

#set domain
nx=32
nz=30
L=1*10**-2
H=1*10**-3
grid=CartesianGrid([[-L,L],[-L,L],[-H,H]],[nx,nx,nz],periodic=[False, False,False])

#This is my question, Can I register a internal heat scalarfield into my PDE? The following code won't work but just let you know what I'm trying to do.

#setup a scalarfield for internal hit source, let's say we have a internal heat surface in the middle of domain
heat=np.ones([32,32,30])*0
heat[:,:,15]=10
heat=ScalarField(grid,ini)

#Try to register C as an new operator. This doesn't work
CartesianGridBase.register_operator('C', heat)

#set up PDE

eq=PDE({"T": f"gradient_x(0.00001*gradient_x(T))+gradient_y(0.00001*gradient_y(T))+gradient_z(0.0000005*gradient_z(T))+C"}, bc={"value": 0})

#setup domain and IC
ini=np.ones([32,32,30])*0
ini[:,:,15]=1
field=ScalarField(grid,ini)

#solve it
storage = MemoryStorage()
res = eq.solve(field, 0.001, dt=5e-7, tracker=storage.tracker(0.0001))
```

**Reply by david-zwicker** [reply]
A few comments:

- Instead of using `gradient_x` you can simply write `d_dx` and you don't need to register the operator beforehand.
- You can use constant fields in the expressions. You simply need to pass them using the `consts` argument of the `PDE` class.

**Reply by louis10643** [reply]
Thank you for your quick reply again! I used d_dx, d_dy, d_dz earlier but it shows errors. This is my PDE:
 
eq=PDE({"T": d_dx(0.00001*d_dx(T))+d_dy(0.00001*d_dy(T))+d_dz(0.0000005*d_dz(T))+C"}, bc={"value": 0})

**Reply by david-zwicker** [reply]
You need multiplication between numbers and operators in python/sympy.

**Reply by louis10643** [reply]
I see, thank you!

**Reply by david-zwicker** [reply]
If this doesn't work, it would be good to open a bug report.
